### PR TITLE
Add haproxy ingress controller

### PR DESCRIPTION
**Pull request by xianlubird**
**What this PR does / why we need it**:

Add haproxy ingress controller

**Special notes for your reviewer**:

### CODE REVIEW

File: incubator/haproxy-ingress/templates/controller-service.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      nodePort: {{ .Values.controller.service.nodePorts.http }}
+      {{- end }}
+    - name: https
+      port: 443
+      protocol: TCP
+      targetPort: {{ .Values.controller.service.targetPorts.https }}
+      {{- if (and (eq .Values.controller.service.type "NodePort") (not (empty .Values.controller.service.nodePorts.https))) }}
+      nodePort: {{ .Values.controller.service.nodePorts.https }}
+      {{- end }}
+  selector:
```
Comment (cheyang) [inline incubator/haproxy-ingress/templates/controller-service.yaml:35]: Please refer [names-and-labels](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#names-and-labels) to set selector.

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+  labels:
```
Comment (cheyang) [inline incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Please refer [names-and-labels](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#names-and-labels) to set labels.

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      run: {{ template "haproxy-ingress.name" . }}
+  template:
+    metadata:
+      labels:
+        run: {{ template "haproxy-ingress.name" . }}
+    spec:
+      serviceAccountName: {{ if or .Values.rbac.create .Values.rbac.createRole .Values.rbac.createClusterRole }}{{ template "haproxy-ingress.fullname" . }}{{ else }}"{{ .Values.rbac.serviceAccountName }}"{{ end }}
+      containers:
+      - name: {{ template "haproxy-ingress.controller.fullname" . }}
+        image:  {{ .Values.controller.image }}
```
Comment (cheyang) [inline incubator/haproxy-ingress/templates/controller-daemonset.yaml]: You'd better to use image repository and tag to specify image
```
image:
  repository: myapp
  tag: 1.2.3
  pullPolicy: IfNotPresent
```

File: incubator/haproxy-ingress/values.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# Use this namespace to install controller
+namespace: ingress-controller-haproxy
+
+# Enable RBAC
+rbac:
+  create: true
```
Comment (cheyang) [inline incubator/haproxy-ingress/values.yaml]: Why do we need `create: true`?

File: incubator/haproxy-ingress/templates/namespace.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
Comment (cheyang) [inline incubator/haproxy-ingress/templates/namespace.yaml]: Do we have to use `namespace` resource? Can we use 'helm install --namespace myns mychart' ?

**Review by cheyang — CHANGES_REQUESTED** [review]
Please read https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md carefully and update according its requirement. Thanks.

File: incubator/haproxy-ingress/values.yaml (hunk 4 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/values.yaml]: Sync to https://github.com/kubernetes/charts/blob/master/stable/nginx-ingress/values.yaml#L334

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml (hunk 2 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Done

File: incubator/haproxy-ingress/templates/controller-service.yaml (hunk 1 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/controller-service.yaml:35]: Done

**Comment by xianlubird** [comment]
ping @mattfarina @scottrigby @unguiculus @foxish @viglesiasce @seanknox @sameersbn @prydonius @mgoodness @linki @lachie83

**Comment by xianlubird** [comment]
any update @mattfarina @scottrigby @unguiculus @foxish @viglesiasce @seanknox @sameersbn @prydonius @mgoodness @linki @lachie83

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Use apps/v1beta2

File: incubator/haproxy-ingress/README.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
+# haproxy-ingress
+
+[haproxy-ingress](https://github.com/jcmoraisjr/haproxy-ingress) is an Ingress controller that uses ConfigMap to store the haproxy configuration.
+
+## Introduction
+
+This chart bootstraps an haproxy-ingress deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+## Prerequisites
+  - Kubernetes 1.6+ with Beta APIs enabled
```
Comment (unguiculus) [inline incubator/haproxy-ingress/README.md]: 1.8+

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+kind: DaemonSet
+metadata:
+  labels:
+    app: {{ template "haproxy-ingress.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    component: "{{ .Values.controller.name }}"
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "haproxy-ingress.controller.fullname" . }}
+  namespace: {{ .Release.Namespace }}
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Remove namespaces from resources. It's unnecessary.

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  template:
+    metadata:
+      labels:
+        app: {{ template "haproxy-ingress.name" . }}
+        component: "{{ .Values.controller.name }}"
+        release: {{ .Release.Name }}
+    spec:
+      serviceAccountName: {{ if or .Values.rbac.create .Values.rbac.createRole .Values.rbac.createClusterRole }}{{ template "haproxy-ingress.fullname" . }}{{ else }}"{{ .Values.rbac.serviceAccountName }}"{{ end }}
+      containers:
+      - name: {{ template "haproxy-ingress.controller.fullname" . }}
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/controller-daemonset.yaml]: I'd like you to make list indentation consistent throughout the chart. Either use zero or two spaces.

File: incubator/haproxy-ingress/templates/default-backend-deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/default-backend-deployment.yaml]: apps/v1beta2

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "haproxy-ingress.controller.fullname" . }}
+  namespace: {{ .Release.Namespace }}
+spec:
+  updateStrategy:
+    type: RollingUpdate
+  selector:
+    matchLabels:
+      app: {{ template "haproxy-ingress.name" . }}
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/controller-daemonset.yaml:15]: Add release and component label to selector.

File: incubator/haproxy-ingress/templates/default-backend-deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    app: {{ template "haproxy-ingress.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    component: "{{ .Values.defaultBackend.name }}"
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ .Values.defaultBackend.name }}
+  namespace: {{ .Release.Namespace }}
+spec:
+  selector:
+    matchLabels:
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/default-backend-deployment.yaml:13]: Add release and component label to selector.

**Review by unguiculus — CHANGES_REQUESTED** [review]
I made an initial review. Please make sure you follow our best practices regarding resource names and labels.

https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Also, please apply our RBAC best practices.

https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml (hunk 11 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/controller-daemonset.yaml:15]: Done

File: incubator/haproxy-ingress/templates/default-backend-deployment.yaml (hunk 12 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/default-backend-deployment.yaml:13]: Done

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml (hunk 6 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Done

File: incubator/haproxy-ingress/README.md (hunk 7 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/README.md]: Done

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml (hunk 8 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Done

File: incubator/haproxy-ingress/templates/controller-daemonset.yaml (hunk 9 above)
Comment (xianlubird) [reply incubator/haproxy-ingress/templates/controller-daemonset.yaml]: Done

**Comment by xianlubird** [comment]
@unguiculus Thanks for your review. Already change the code. Pls review again if you are free, thanks.
 any update @unguiculus 
.....

**Comment by xianlubird** [comment]
any update @mattfarina @scottrigby @unguiculus @foxish @viglesiasce @seanknox @sameersbn @prydonius @mgoodness @linki @lachie83 ....

**Comment by unguiculus** [comment]
Please have a look at our RBAC best practices again. Specifically, you should create a partial for the service account.

**Comment by unguiculus** [comment]
And please don't squash or rebase, so Github can show your changes.

**Comment by xianlubird** [comment]
@unguiculus Already change serviceAccount section, many thanks.

**Comment by xianlubird** [comment]
@unguiculus Change some code according to your PR  #5639  :)

**Comment by xianlubird** [comment]
any update @mattfarina @scottrigby @unguiculus @foxish @viglesiasce @seanknox @sameersbn @prydonius @mgoodness @linki @lachie83 ....

**Comment by xianlubird** [comment]
any update  @scottrigby @unguiculus @foxish @viglesiasce @seanknox @sameersbn @prydonius @mgoodness @linki @lachie83 @mattfarina

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please compare with what you get from `helm create` using Helm 2.8+ and update accordingly.

**Comment by xianlubird** [comment]
@unguiculus Thanks for your review.  Already change the code.

File: incubator/haproxy-ingress/templates/_helpers.tpl [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+{{- end -}}
+{{- end -}}
+
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+If release name contains chart name it will be used as a full name.
+*/}}
+{{- define "haproxy-ingress.controller.fullname" -}}
```
Comment (unguiculus) [inline incubator/haproxy-ingress/templates/_helpers.tpl]: This seems to do exactly the same as `haproxy-ingress.fullname`. So, please remove it and update references.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by xianlubird** [comment]
@unguiculus Thanks for your review. Already remove the `haproxy-ingress.controller.fullname`

**Review by unguiculus — APPROVED** [review]

**Pull request merged**